package/network/services/openvpn/files/etc/hotplug.d/openvpn/01-user

```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
+
+[ -e "/etc/openvpn.user" ] && {
+	env -i ACTION="$ACTION" INSTANCE="$INSTANCE" \
+		/bin/sh \
+		/etc/openvpn.user \
+		$*
+}
+
+# Wrap user defined scripts on up/down events
+. /lib/functions.sh
+config_load 'openvpn'
+LOG="logger -t openvpn/hotplug[$$] -p"
+
+[ "$ACTION" = "up" ] && {
+	config_get up "$INSTANCE" up
+	[ -z "$up" ] || {
+		if [ -f "$up" ]; then
+			/bin/sh "$up" \
```

Comment:
Please use `[ -n "$up" ]` and `/bin/sh -c "$up" ...`.

The OpenVPN help states:
> --up cmd
> Shell  command  to run after successful TUN/TAP device open

So that seems to imply that both something like `echo 1` or `/usr/bin/example` is supposed to work.

Reply:
> Please use [ -n "$up" ] 

OK will change this

> So that seems to imply that both something like echo 1 or /usr/bin/example is supposed to work.

Yes this is right but I get the following error if I add `echo 1` to the **down** option of a openvpn uci config section. From my point of view it does not make sense to only add a command line string,  because you have to wrap all this quotation `' " `in the uci config. I think it would be better for the user to add their command into a script.
The advantage is he has all the possibilities a shell offers and he doesn't have to be annoyed with quotation marks.

```
config openvpn 'test'
     option down 'echo 1'
```
> Options error: the --down directive should have at most 1 parameter.  To pass a list of arguments as one of the parameters, try enclosing them in double quotes ("").
> Thu Apr  4 11:11:04 2019 daemon.warn openvpn(test)[20360]: Use --help for more information.

Reply:
I have update the check but I think that using only a command and not a script is not the way to go. For the reasons given above.

Reply:
It still requires the command to be a file. Command strings like `route add ... && route add ... && /etc/init.d/foo restart` are perfectly valid.

Reply:
You also forgot to pass `-c` to `/bin/sh`. Without that, the shell will attempt to parse the argument as script, which will not work for ELF executables:

```
root@jj:~# /bin/sh /bin/date
/bin/date: line 2: syntax error: unterminated quoted string
root@jj:~# /bin/sh -c /bin/date
Wed Apr 24 11:03:30 CEST 2019
```

Reply:
Sorry I do not get it.
If I set `--up "/usr/bin/logger hello1 && /usr/bin/logger hello"` only the first logger is called 

> openvpn(test)[3023]: /usr/bin/logger hello1 && /usr/bin/logger hello tun0 1500 1544 10.9.8.2 10.9.8.1 init
>  user.notice root: hello1 && /usr/bin/logger hello tun0 1500 1544 10.9.8.2 10.9.8.1 init

Reply:
The following works in native configuration:

    up '/bin/sh -c "/usr/bin/logger test1 && /usr/bin/logger test2"'

The following works in UCI:

    option up "\"/bin/sh -c '/usr/bin/logger test1 && /usr/bin/logger test2'\""

Your check for `[ -f "$up" ]` breaks these use cases and forcing the command string through `/bin/sh` without `-c` breaks normal non-shellscript programs as well.

You may not make assumptions about the contents of `$up` and `$down` other than that they're  command vectors to be passed to `execve()`.  OpenVPN actually tokenizes the `up` or `down` values into lists while respecting quotes and escapes. So the example values above `/bin/sh -c "/usr/bin/logger test1 && /usr/bin/logger test2"` becomes `{ "/bin/sh", "-c", "/usr/bin/logger test1 && /usr/bin/logger test2" }` which is then passed to `execve()` as follows: `execve("/bin/sh", { "/bin/sh", "-c", "/usr/bin/logger test1 && /usr/bin/logger test2", "hello", "tun0", "1500", "1544", "10.9.8.2", "10.9.8.1", "init" }, envptr)`.

A simpler, more real word example would be something similar to `up "/usr/local/bin/my-custom-handler --action=init --instance=foo --"` which will get passed as `execve("/usr/local/bin/my-custom-handler", { "/usr/local/bin/my-custom-handler", "--action=init", "--instance=foo", "--", "hello", "tun0", "1500", "1544", "10.9.8.2", "10.9.8.1", "init" }, envptr)`.

So in short, testing for `[ -f "$up" ]` will break the ability to specify commands with arguments (which is supported by OpenVPN) and running through `/bin/sh` instead of `/bin/sh -c` will break all variants where the value refers to a path that is not a shell script.

Reply:
@jow Thank you very much or the clarifying words. I have now added your requested changes. I have also remove **up**, **down** and **script_security** from the openvpn.options file. If I do not do this i will get the following erros during service start

> daemon.err openvpn(test)[2684]: Options error: the --down directive should have at most 1 parameter.  To pass a list of arguments as one of the parameters, try enclosing them in double quotes ("").

This is because the command will also get moved to the openvpn.config but I think that's wrong., because the options are now added during openvpn start. This must also be true for the remaining parameters **cd**, **status** and **syslog**. The option **config** is already not included in this file.